core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -165,6 +165,26 @@ final case class NonEmptyList[A](head: A, tail: List[A]) {
     go(head, tail, Nil)
   }
 
+  /**
+   * Zips each element of this `NonEmptyList` with its index.
+   *
+   * {{{
+   * scala> import cats.data.NonEmptyList
+   * scala> val nel = NonEmptyList.of("a", "b", "c")
+   * scala> nel.zipWithIndex
+   * res0: cats.data.NonEmptyList[(String, Int)] = NonEmptyList((a,0), (b,1), (c,2))
+   * }}}
+   */
+  def zipWithIndex: NonEmptyList[(A, Int)] = {
+    @tailrec
+    def go(rest: List[A], ix: Int, acc: List[(A, Int)]): List[(A, Int)] =
```

Comment:
while this is a nice implementation, I think `ListBuilder` can break the rules and mutate `List`s and be faster. So, something like;
```scala
val bldr = List.newBuilder[(A, Int)]
var idx = 1
val it = tail.iterator
while (it.hasNext) {
  bldr += (it.next, idx)
  idx += 1
}
NonEmptyList((head, 0), bldr.result)
```
will be faster, which is nice when we are talking about common libraries.